[RDY] vim-patch:8.0.0642

janlazo opened the pull request:
**vim-patch:8.0.0642: writefile() continues after detecting an error**

Problem:    writefile() continues after detecting an error.
Solution:   Bail out as soon as an error is detected. (suggestions by Nikolai
            Pavlov, closes vim/vim#1476)
https://github.com/vim/vim/commit/8cf91286ca46a501d24e4b7d631b193256782c88

janlazo committed:
vim-patch:8.0.0642: writefile() continues after detecting an error

Problem:    writefile() continues after detecting an error.
Solution:   Bail out as soon as an error is detected. (suggestions by Nikolai
            Pavlov, closes vim/vim#1476)
https://github.com/vim/vim/commit/8cf91286ca46a501d24e4b7d631b193256782c88

janlazo commented:
```
[  FAILED  ] ...s/build/neovim/neovim/test/functional/eval/null_spec.lua @ 29: NULL list is accepted as an empty list by writefile()
...s/build/neovim/neovim/test/functional/eval/null_spec.lua:30: Expected objects to be the same.
Passed in:
(string) '
E484: Can't open file Xtest-functional-viml-null'
Expected:
(string) ''
stack traceback:
	...s/build/neovim/neovim/test/functional/eval/null_spec.lua:30: in function <...s/build/neovim/neovim/test/functional/eval/null_spec.lua:29>
[  FAILED  ] ...ld/neovim/neovim/test/functional/eval/writefile_spec.lua @ 130: writefile() stops writing to file after error in list
...ld/neovim/neovim/test/functional/eval/writefile_spec.lua:134: Expected objects to be the same.
Passed in:
(nil)
Expected:
(string) 'tset
'
stack traceback:
	...ld/neovim/neovim/test/functional/eval/writefile_spec.lua:134: in function <...ld/neovim/neovim/test/functional/eval/writefile_spec.lua:130>
```

justinmk commented:
https://github.com/neovim/neovim/pull/6111 is related, but commit message in https://github.com/ZyX-I/neovim/commit/f489827b5f9b8b1ecd684768a92871389217aa91 suggests this was a TODO. So we do need this vim patch.

justinmk commented on src/nvim/eval.c:
```diff
@@ -17685,6 +17685,15 @@ static void f_writefile(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  const list_T *const list = argvars[0].vval.v_list;
+  if (list == NULL) {
+    return;
+  }
```
This change seems to contradict the existing test `NULL list is accepted as an empty list by writefile()`. 
I think we can leave this out, because we don't consider it an error. cc @ZyX-I

ZyX-I replied:
NULL list should still create a file, it is incorrect to exit here.

justinmk commented:
https://github.com/neovim/neovim/pull/6111 is related, the commit message in https://github.com/ZyX-I/neovim/commit/f489827b5f9b8b1ecd684768a92871389217aa91 mentions the need for `tv_check_str_or_nr`. That was merged.

So I think we don't need the code changes. Only the vim-style test, which should pass without changes.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -17685,6 +17685,15 @@ static void f_writefile(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+    return;
+  }
+  TV_LIST_ITER_CONST(list, li, {
+    if (tv_get_string_chk(TV_LIST_ITEM_TV(li)) == NULL) {
```
There is `tv_check_str_or_nr()` function now, it should be used here in place of `tv_get_string_chk()` which will do useless job of formatting number.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -17685,6 +17685,15 @@ static void f_writefile(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  if (list == NULL) {
+    return;
+  }
+  TV_LIST_ITER_CONST(list, li, {
```
TV_LIST_ITER_CONST is perfectly fine with a NULL list. Won’t execute a list body even once obviously, but still fine.

ZyX-I replied:
So you should be able to just remove that `if()` block and everything would work as intended.

janlazo committed:
test: update writefile test for invalid list items

It validates list items with tv_check_str_on_nr()
to catch invalid types (ex. E745, E805).
If there is an invalid item, it does not write to the file.

janlazo commented:
Lint warnings from https://github.com/neovim/neovim/pull/8623 caused the lint build to failed.

janlazo commented:
Any issues with the new behaviour for `writefile`?

justinmk commented on test/functional/eval/writefile_spec.lua:
```diff
@@ -129,21 +129,21 @@ describe('writefile()', function()
+    eq('\nE745: Expected a Number or a String, List found',
         redir_exec(('call writefile(%s)'):format(args:format('[]'))))
-    eq('tset\n', read_file(fname))
+    eq('TEST', read_file(fname))
```
It seems that this test group `it('stops writing to file after error in list'` isn't really being tested meaningfully anymore. Because the write doesn't begin at all. 

    it('does not write to file if error in list', function()

cc @ZyX-I
